frame/session/src/historical/storage.rs

```diff
@@ -0,0 +1,138 @@
+//!
+//! This is used in conjunction with [`ProvingTrie`](super::ProvingTrie) and
+//! the offchain indexing API.
+
+use sp_io::offchain_index;
+use sp_std::prelude::*;
+
+use super::super::{Module as SessionModule, SessionIndex};
+use super::Trait;
+
+const PREFIX: &[u8] = b"historical";
+const LAST_PRUNE: &[u8] = b"historical_last_prune";
+const HEAD: &[u8] = b"historical_head";
+
+pub struct ValidatorSet<T: Trait> {
+    validator_set: Vec<(T::ValidatorId, T::FullIdentification)>,
+}
+
+/// Derive the key used to store the list of validators
+fn derive_key<P: AsRef<[u8]>, T: Trait>(prefix: P, session_index: Vec<u8>) -> Vec<u8> {
+    assert!(session_index.len() > 0);
+    let prefix = prefix.as_ref();
+    let mut concatenated = Vec::with_capacity(prefix.len() + 1 + session_index.len());
+    concatenated.extend_from_slice(prefix);
+    concatenated.push('/');
+    (&mut concatenated[(prefix.len() + 1)..]).extend_from_slice(session_index.as_slice());
+    concatenated
+}
+
+impl<T: Trait> ValidatorSet<T> {
+    /// Load the set of validators for a paritcular session index from the offchain storage.
+    ///
+    /// If none is found or decodable given `prefix` and `session`, it will return `None`.
+    /// Empty validator sets should only ever exist for genesis blocks.
+    fn load_from_offchain(session_index: SessionIndex) -> Option<Self> {
+        let derived_key = derive_key(STATIC_PREFIX, session_index.encode());
+        let validator_set = StorageValueRef::persistent(derived_key.as_ref())
+            .get::<Vec<(T::ValidatorId, T::FullIdentification)>>()
+            .flatten();
+        validator_set
+    }
+
+    /// Access the underlying `ValidatorId` and `FullIdentification` tuples as slice.
+    fn as_slice(&self) -> &[(T::ValidatorId, T::FullIdentification)] {
+        self.validator_set.as_slice()
+    }
+
+    /// Convert `self` to a vector and consume `self`.
+    fn to_vec(self) -> Vec<(T::ValidatorId, T::FullIdentification)> {
+        self.validator_set
+    }
+
+    /// Prune anything older than the current session index.
+    ///
+    /// For behaviour details refer to [`fn prune_older_than`](Self::prune_older_than).
+    ///
+    /// **Must** be called from the offchain worker.
+    fn prune() {
+        let move_pruning_marker = SessionModule::current_index();
+        Self::prune_older_than(move_pruning_marker);
+    }
+
+    /// Attempt to prune anything that is older than `first_to_keep` session index.
+    ///
+    /// Due to re-ogranisation it could be that the `first_to_keep` might be less
+    /// than the stored one, in which case the conservative choice is made to keep records
+    /// up to the one that is the lesser.
+    ///
+    /// **Must** be called from the offchain worker.
+    fn prune_older_than(first_to_keep: SessionIndex) {
+        let pruning_marker_key = derive_key(STATIC_LAST_PRUNE);
+        let mut entry = StorageValueRef::persistent(derived_key.as_ref());
```

Comment:
This function probably does not compile. But it's meant to be `pruning_marker_key` instead of `derived_key` here, right?

Reply:
Yep, backup squash commit & push, will fix that asap